llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,352 @@
+      !TII->isHardwareLoopEnd((*Terminator).getOpcode())) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No ZOL\n");
+    return false;
+  }
+  // 2. We need to fix the tripcount and to push out a time-critical prologue.
+  // If we don't have a dedicated preheader that is fallthrough, don't even
+  // bother.
+  Preheader = AIELoopUtils::getDedicatedFallThroughPreheader(LoopBlock);
+  if (!Preheader) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No fallthrough preheader\n");
+    return false;
+  }
+
+  // 3. We need to know how to update the tripcount. We check whether
+  // the tripcount is pristine, otherwise the loop may have been software
+  // pipelined before and we can't trust min itercount metadata.
+  // Return on investment is probably low anyway.
+  const bool Pristine = true;
+  for (auto &MI : reverse(*Preheader)) {
+    if (TII->isZOLTripCountDef(MI, Pristine)) {
+      TripCountDef = &MI;
+      break;
+    }
+  }
+  if (!TripCountDef) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No tripcount def\n");
+    return false;
+  }
+
+  // 4. We need to peel stages and be left with a positive tripcount.
+  // This is just a minimum check to save useless work; the real stage
+  // count is checked before accepting the schedule.
+  auto ParsedMinTripCount = AIELoopUtils::getMinTripCount(LoopBlock);
+  if (!ParsedMinTripCount) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No min tripcount\n");
+    return false;
+  }
+  MinTripCount = *ParsedMinTripCount;
+  if (MinTripCount < 2) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: min tripcount < 2\n");
+    return false;
+  }
+
+  return true;
+}
+
+// This assigns Cycle of SU, Earliest of its predecessors and Earliest of
+// the next instance of SU.
+void PostPipeliner::scheduleNode(SUnit &SU, int Cycle) {
+  LLVM_DEBUG(dbgs() << "PostPipeline " << SU.NodeNum << " in cycle " << Cycle
+                    << "\n");
+  Info[SU.NodeNum].Cycle = Cycle;
+  for (auto &Dep : SU.Succs) {
+    int Latency = Dep.getSignedLatency();
+    auto *Succ = Dep.getSUnit();
+    if (Succ->isBoundaryNode()) {
+      continue;
+    }
+    int SNum = Succ->NodeNum;
+    Info[SNum].Earliest = std::max(Info[SNum].Earliest, Cycle + Latency);
+    LLVM_DEBUG(dbgs() << "   Earliest(" << SNum << ") to "
+                      << Info[SNum].Earliest << "\n");
+  }
+
+  int Next = SU.NodeNum + NInstr;
+  if (Next < NTotalInstrs) {
+    Info[Next].Earliest = std::max(Info[Next].Earliest, Cycle + II);
+  }
+}
+
+// Check resources. We only insert at the position modulo II. Since we insert
+// all iterations separately, the resources that wrap around accumulate in the
+// overflow area, causing conflicts when inserting future iterations
+int PostPipeliner::fit(MachineInstr *MI, int Earliest, int NTries, int II) {
+  for (int C = Earliest; C < Earliest + NTries; C++) {
+    int Mod = C % II;
+    LLVM_DEBUG(dbgs() << "   at " << C << " (" << Mod << ")\n");
+    if (!HR.checkConflict(Scoreboard, *MI, -Depth + Mod)) {
```

Comment:
Is that really enough to verify conflicts at `-Depth + Mod + II`, `-Depth + Mod + 2*II`, etc.? Or do we have a special scoreboard here that has a size of `II` and wraps around?

Reply:
Hmm. I think I have explained in a number of places that the +II +2*II etc conflicts are accumulating by scheduling all iterations into steady state.

Reply:
Yes I just read the rest of the code now. Coming back here, I can understand it is not needed. There's just a lot of changes to process at once :)